File: 2/js/src/SmallestDifference.ts
function getSmallestDifference(list1: number[], list2: number[]): [number, number] {
	list1.sort((a,b) => a-b);
	list2.sort((a,b) => a-b);

	let res: [number, number] = [0, Number.MIN_SAFE_INTEGER];
	let ptr1 = 0;
	let ptr2 = 0;

	while(ptr1 < list1.length && ptr2 < list2.length) {
		const x = list1[ptr1];
		const y = list2[ptr2];

		const diff = Math.abs(x - y);
		if(diff <= Math.abs(res[1] - res[0]))	{
			res[0] = x;
			res[1] = y;
		}

		if(x < y) {
			ptr1++;
		} else if(x > y) {
			ptr2++;
		} else {
			return [x,y];
		}
	}

	return res;
}

(function main() {
	const list1 = [-1,5,10,20,28,3];
	const list2 = [26,134,135,15,17];

	console.log(getSmallestDifference(list1, list2));
})();

File: 2/java/SmallestDifference.java
import java.util.Arrays;

class SmallestDifference {
	public static int[] getSmallestDifference(int array1[], int array2[]) {
		int diff = Integer.MAX_VALUE;
		int res[] = new int[2];

		Arrays.sort(array1);
		Arrays.sort(array2);
		
		int i = 0;
		int j = 0;

		while(i < array1.length && j < array2.length) {
			int num1 = array1[i];
			int num2 = array2[j];

			int newDiff = Math.abs(num1-num2);

			if(num1 < num2) {
				i++;
			}
			else if(num1 > num2) {
				j++;
			} else {
				return new int[]{num1, num2};
			}
			
			if(newDiff < diff) {
				diff = newDiff;

				res[0] = num1;
				res[1] = num2;
			}
		}
		
		return res;
	}
	public static void main(String args[]) {
		int array1[] = new int[]{-1,5,10,20,28,3};
		int array2[] = new int[]{26,134,135,15,17};
		
		int res[] = getSmallestDifference(array1, array2);

		System.out.println(String.format("%d %d", res[0], res[1]));

	}
}
